File: backend/src/controllers/events.controller.ts
import { Request, Response } from 'express'
import { supabase } from '../config/supabase'

// Lista todos os eventos do usuário autenticado
export async function listarEventos(req: Request, res: Response) {
  const { data, error } = await supabase
    .from('events')
    .select('*')
    .eq('user_id', req.user!.id)
    .order('date', { ascending: true })

  if (error) return res.status(500).json({ error: error.message })
  res.json(data)
}

// Cria um novo evento
export async function criarEvento(req: Request, res: Response) {
  const { title, date, type, description } = req.body

  if (!title?.trim() || !date) {
    return res.status(400).json({ error: 'Título e data são obrigatórios' })
  }

  const { data, error } = await supabase
    .from('events')
    .insert({
      user_id: req.user!.id,
      title: title.trim(),
      date,
      type: type || 'Pessoal',
      description: description?.trim() || null,
    })
    .select()
    .single()

  if (error) return res.status(500).json({ error: error.message })
  res.status(201).json(data)
}

// Atualiza um evento existente
export async function editarEvento(req: Request, res: Response) {
  const { id } = req.params
  const { title, date, type, description } = req.body

  const { data, error } = await supabase
    .from('events')
    .update({ title, date, type, description })
    .eq('id', id)
    .eq('user_id', req.user!.id)
    .select()
    .single()

  if (error) return res.status(500).json({ error: error.message })
  if (!data) return res.status(404).json({ error: 'Evento não encontrado' })
  res.json(data)
}

// Remove um evento
export async function deletarEvento(req: Request, res: Response) {
  const { id } = req.params

  const { error } = await supabase
    .from('events')
    .delete()
    .eq('id', id)
    .eq('user_id', req.user!.id)

  if (error) return res.status(500).json({ error: error.message })
  res.json({ ok: true })
}

File: backend-go/internal/events/events_controller.go
package events

import (
	"net/http"
	"strconv"
	"strings"

	"github.com/gin-gonic/gin"
	"github.com/jackc/pgx/v5"
	"github.com/jackc/pgx/v5/pgxpool"

	"beeplanerofc/backend-go/internal/dbutil"
	"beeplanerofc/backend-go/internal/middleware"
)

type Controller struct {
	db *pgxpool.Pool
}

func NewController(db *pgxpool.Pool) *Controller {
	return &Controller{db: db}
}

// GET /api/events — lista os eventos do usuário, ordenados por data
func (ctrl *Controller) ListarEventos(c *gin.Context) {
	user := middleware.CurrentUser(c)

	rows, err := ctrl.db.Query(c.Request.Context(),
		"SELECT * FROM events WHERE user_id = $1 ORDER BY date ASC", user.ID,
	)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
		return
	}

	eventos, err := pgx.CollectRows(rows, dbutil.RowToMap)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
		return
	}

	c.JSON(http.StatusOK, eventos)
}

type criarEventoRequest struct {
	Title       string  `json:"title"`
	Date        string  `json:"date"`
	Type        *string `json:"type"`
	Description *string `json:"description"`
}

// POST /api/events — cria um novo evento
func (ctrl *Controller) CriarEvento(c *gin.Context) {
	user := middleware.CurrentUser(c)

	var req criarEventoRequest
	c.ShouldBindJSON(&req)

	title := strings.TrimSpace(req.Title)
	if title == "" || req.Date == "" {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Título e data são obrigatórios"})
		return
	}

	eventType := "Pessoal"
	if req.Type != nil && *req.Type != "" {
		eventType = *req.Type
	}

	var description *string
	if req.Description != nil {
		trimmed := strings.TrimSpace(*req.Description)
		if trimmed != "" {
			description = &trimmed
		}
	}

	rows, err := ctrl.db.Query(c.Request.Context(), `
		INSERT INTO events (user_id, title, date, type, description)
		VALUES ($1, $2, $3, $4, $5)
		RETURNING *`,
		user.ID, title, req.Date, eventType, description,
	)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
		return
	}

	evento, err := pgx.CollectOneRow(rows, dbutil.RowToMap)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
		return
	}

	c.JSON(http.StatusCreated, evento)
}

type editarEventoRequest struct {
	Title       *string `json:"title"`
	Date        *string `json:"date"`
	Type        *string `json:"type"`
	Description *string `json:"description"`
}

// PUT /api/events/:id — atualiza os campos enviados de um evento
func (ctrl *Controller) EditarEvento(c *gin.Context) {
	user := middleware.CurrentUser(c)
	id := c.Param("id")

	var req editarEventoRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Dados inválidos"})
		return
	}

	setClauses := []string{}
	args := []any{}
	addField := func(column string, value any) {
		args = append(args, value)
		setClauses = append(setClauses, column+" = $"+strconv.Itoa(len(args)))
	}

	if req.Title != nil {
		addField("title", *req.Title)
	}
	if req.Date != nil {
		addField("date", *req.Date)
	}
	if req.Type != nil {
		addField("type", *req.Type)
	}
	if req.Description != nil {
		addField("description", *req.Description)
	}

	if len(setClauses) == 0 {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Nenhum campo para atualizar"})
		return
	}

	args = append(args, id, user.ID)
	query := "UPDATE events SET " + strings.Join(setClauses, ", ") +
		" WHERE id = $" + strconv.Itoa(len(args)-1) + " AND user_id = $" + strconv.Itoa(len(args)) +
		" RETURNING *"

	rows, err := ctrl.db.Query(c.Request.Context(), query, args...)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
		return
	}

	evento, err := pgx.CollectOneRow(rows, dbutil.RowToMap)
	if err == pgx.ErrNoRows {
		c.JSON(http.StatusNotFound, gin.H{"error": "Evento não encontrado"})
		return
	}
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
		return
	}

	c.JSON(http.StatusOK, evento)
}

// DELETE /api/events/:id — remove um evento
func (ctrl *Controller) DeletarEvento(c *gin.Context) {
	user := middleware.CurrentUser(c)
	id := c.Param("id")

	_, err := ctrl.db.Exec(c.Request.Context(),
		"DELETE FROM events WHERE id = $1 AND user_id = $2", id, user.ID,
	)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
		return
	}

	c.JSON(http.StatusOK, gin.H{"ok": true})
}
